Add instance-type node label to cloud providers

therc opened the pull request:
Implemented for AWS and GCE. Tested on AWS. See #9044 

I looked around and couldn't find if there's anything that can be done in practice for the other providers.

Use cases: schedule on GPU instances or those with better I/O guarantees, etc. Or prevent scheduling on burstable instances.

It would be nice if, instead of raw instance names, we had a way to expose instance traits (burstable, etc.) but e.g. the EC2 API has no way to discover programmatically the instance types available in a region, let alone their characteristics.

Commit:
Add instance-type label to cloud providers

Fully implemented for AWS and GCE

therc commented:
I added logging like in another couple of spots later in the code. Not sure why the e2e build failed, hoping it was a flake.

therc commented:
Probably tripped over getInstanceByNames(). Teaches me about trying to implement a feature for a system I can't test on. :-) Let's see if third time's a charm.

therc commented:
I'm at a loss. How to debug this? I can't get to the logs from the nodes or the master.

Or maybe I should just take the gce code out.

therc commented:
Cc @justinsb for the AWS part

therc commented:
@bgrant0607 would you be OK with this change? I only want to schedule some pods on better hardware. Or avoid Noconas. :-)

bgrant0607 commented:
cc @davidopp @vishh

bgrant0607 commented on pkg/cloudprovider/providers/aws/aws.go:
```diff
@@ -740,6 +744,24 @@ func (aws *AWSCloud) InstanceID(name string) (string, error) {
 
+// InstanceType returns the type of the specified instance.
+func (aws *AWSCloud) InstanceType(name string) (string, error) {
+	awsInstance, err := aws.getSelfAWSInstance()
```
Could this use metadata.GetMetadata?

therc replied:
I'm following the pattern from other functions, such as InstanceID(). They all let you call them from either the node itself or from another (e.g. a controller). metadata.GetMetadata() is local only.

bgrant0607 commented:
I'm fine with the idea of an instance-type label.

davidopp commented:
I'm holding off on this pending the discussion in #21024

therc commented:
Hah! I was looking forward to a reviewer's guidance for the label, as I had quickly made it up without knowing if it'd pass muster. I'll follow the other discussion, too.

davidopp commented:
Looks like the decision is to use "beta" rather than "alpha". Please update the PR.

davidopp commented:
@bgrant0607 is it OK to have a label key with nothing before the alpha.kubernetes.io (or beta.kubernetes.io) ?

therc commented:
Updated. I'll name it Woger or Woderick, if I have to. :-)

therc commented:
I'm assuming that the e2e test was failing because in getMachinesByNames there was a filtered List call that asks for an explicit list of fields to be returned. Fixed to also ask to retrieve machineType. And squashed.

justinsb commented on pkg/cloudprovider/providers/aws/aws.go:
```diff
@@ -751,6 +755,24 @@ func (aws *AWSCloud) InstanceID(name string) (string, error) {
+		if err != nil {
+			return "", err
+		}
+		return orEmpty(inst.InstanceType), nil
```
FYI, you can use aws.StringValue instead of orEmpty (I wasn't aware of this, or it wasn't originally available).  No need to change though, just for future reference

justinsb commented:
I'm fine with the AWS side of this

therc commented:
So it just turns out that with GCE the `machineType` field is of the form `zones/us-central1-f/machineTypes/n1-standard-1` and Kubernetes values don't accept slashes, in addition to being limited to 63 characters. Now I'm just grabbing the last component of the path, the same way it's done with the zone field.

davidopp commented:
LGTM

therc commented:
Thanks! Should this be documented or added to release notes? I can send a PR for the former. Where should it be mentioned, though? `docs/user-guide/node-selection` is the first candidate I could find and perhaps it should also talk about the automatic failure domain labels. Where else? `docs/admin/daemons.md`?

davidopp commented:
I'm going to be adding information about our new node affinity stuff to `docs/user-guide/node-selection` so I'll take care of it. Thanks for thinking about documentation!

bgrant0607 commented:
To answer @davidopp's question (belatedly): It is generally a good idea to put something before `beta.kubernetes.io`, to scope the label or annotation further. For instance `node.beta.kubernetes.io` might be appropriate. However, if this case, it's also somewhat redundant, since this is a node label, so the way it is is fine.
